plugins/webp-uploads/tests/test-picture-element.php

```diff
@@ -264,6 +296,110 @@ function ( string $content ): string {
+			foreach ( $custom_sizes as $size ) {
+				$this->temp_custom_image_sizes[] = $size;
+			}
+		}
+
+		update_option( 'perflab_generate_webp_and_jpeg', '1' );
+		update_option( 'perflab_generate_all_fallback_sizes', '1' );
+
+		// Create image with different sizes.
+		$attachment_id = self::factory()->attachment->create_upload_object( TESTS_PLUGIN_DIR . '/tests/data/images/leaves.jpg' );
+		$this->assertNotWPError( $attachment_id );
+		$this->temp_attachment_ids[] = $attachment_id;
+
+		$image = wp_get_attachment_image(
+			$attachment_id,
+			'large',
+			false,
+			array(
+				'class' => 'wp-image-' . $attachment_id,
+				'alt'   => 'Green Leaves',
+			)
+		);
+
+		// Apply picture element support.
+		$this->opt_in_to_picture_element();
+
+		$picture_markup    = apply_filters( 'the_content', $image );
+		$picture_processor = new WP_HTML_Tag_Processor( $picture_markup );
+		$picture_processor->next_tag( array( 'tag_name' => 'SOURCE' ) );
+		$source_srcset = $picture_processor->get_attribute( 'srcset' );
+		$picture_processor->next_tag( array( 'tag_name' => 'IMG' ) );
+		$img_srcset = $picture_processor->get_attribute( 'srcset' );
+
+		// Extract file names from srcset.
+		$extract_file_names_from_srcset = static function ( string $srcset ): array {
+			$srcset_parts = explode( ', ', $srcset );
+			$files        = array();
+			foreach ( $srcset_parts as $srcset_part ) {
+				$parts   = explode( ' ', trim( $srcset_part ) );
+				$url     = $parts[0];
+				$files[] = pathinfo( basename( $url ), PATHINFO_FILENAME );
+			}
+			return $files;
+		};
+
+		$source_files = $extract_file_names_from_srcset( $source_srcset );
+		$img_files    = $extract_file_names_from_srcset( $img_srcset );
+
+		$this->assertCount( count( $source_files ), $img_files );
+
+		foreach ( $img_files as $img_index => $img_file ) {
+			// Check the file name starts with the same prefix.
+			// e.g. $img_file = 'leaves-350x200' and $source_files[ $img_index ] = 'leaves-350x350-jpg'.
+			$this->assertStringStartsWith( $img_file, $source_files[ $img_index ] );
+		}
+	}
+
+	/**
+	 * Data provider for test_picture_element_source_tag_srcset_has_same_image_sizes_as_img_tag_srcset.
+	 *
+	 * @return array<string, array{ set_up: Closure|null, custom_sizes: array<string>|null }>
+	 */
+	public function data_provider_test_picture_element_source_tag_srcset_has_same_image_sizes_as_img_tag_srcset(): array {
+		return array(
+			'default_sizes' => array(
+				'set_up'       => null,
+				'custom_sizes' => null,
+			),
+			'when_two_different_image_sizes_have_same_width' => array(
+				'set_up'       => static function (): void {
+					add_image_size( 'square', 768, 768, true );
+					add_filter(
+						'pre_option_medium_large_size_w',
+						static function () {
+							return '768';
+						}
+					);
+					add_filter(
+						'pre_option_medium_large_size_h',
+						static function () {
+							return '512';
+						}
+					);
+				},
+				'custom_sizes' => array( 'square' ),
```

Comment:
I think we could do without this key, having only `set_up` would be sufficient: You already add the `square` image size in the `set_up` callback. Since the size is otherwise only used to add to `$this->temp_custom_image_sizes`, you could include that in the `set_up` callback instead (change it from `static function ()` to `function ()`).